### ISSUE TITLE
Add STUMPI

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#31]
See a reference implementation of STOMPI in Table 5 in [this paper](https://www.cs.ucr.edu/~eamonn/MP_journal.pdf).

This is the incremental version of STUMP (not interactive!).

[STAMPI](https://www.cs.ucr.edu/~eamonn/PID4481997_extend_Matrix%20Profile_I.pdf) can be found here.

**Comment by seanlaw** [comment]
@rtmatx. This is here for your reference

**Comment by hmcoservit** [comment]
#19 @seanlaw I am new to Matrix profiles and I would like to work on a POC and potentially do a benchmark to demonstrate the usage of MP algorithms for real-time anomaly detection in big data streaming applications. The data source is a number of sensors and at each moment in time 2-3 weeks of recent historical data is available. Depending on the sensor, the data arrives from 1 minute to 1 hours (sampling interval). For this I need to update the matrix profile every time a new data point arrives (forget the oldest point) and decide if a new point is an anomaly or not (using maybe a percentile approach).

From my understanding most of the implementations are adapted to batch based methods (STOMP). I was looking into FLOSS but it is more specific to regimes.  Or [SCRIMP++  ](https://github.com/target/matrixprofile-ts) which is an extension of STOMP (but also for batch). From the article you pointed out, it seems like STUMPI is suitable for such a case, what do you think?

**Comment by seanlaw** [comment]
@hmcoservit Thanks for providing the context and welcome to the STUMPY community! Yes, I think most things sensor/IoT related are certainly within scope and this helps us prioritize our efforts. 

>  I am new to Matrix profiles and I would like to work on a POC and potentially do a benchmark to demonstrate the usage of MP algorithms for real-time anomaly detection in big data streaming applications. The data source is a number of sensors and at each moment in time 2-3 weeks of recent historical data is available. Depending on the sensor, the data arrives from 1 minute to 1 hours (sampling interval). For this I need to update the matrix profile every time a new data point arrives (forget the oldest point) and decide if a new point is an anomaly or not (using maybe a percentile approach).

Can you give me a sense of how many data points you would be analyzing at each iteration? For historical data, you'd be using the batch methods to analyze it anyways but something like STUMPI would be useful for single point updates (essentially, it would be calling the `STAMP` algorithm under the hood). 

Additionally, what kind(s) of hardware resources do you have available to you? GPUs, CPUs?

> From my understanding most of the implementations are adapted to batch based methods (STOMP). I was looking into FLOSS but it is more specific to regimes. Or SCRIMP++ which is an extension of STOMP (but also for batch). From the article you pointed out, it seems like STUMPI is suitable for such a case, what do you think?

Yes, your assessment is correct. But as I mentioned above, you'll be using these batch methods for historical data and calling something like STUMPI just update the matrix profile for the "new" window that is arriving. Have you already thought about what window size, `m`, you'll be using to compute your matrix profile?

**Comment by hmcoservit** [comment]
@seanlaw Thank you for your reply. All of the time series are 1-D and sampled every 10 mins on average (1min-60min). A minimum of 2 weeks of history size is available in memory for each time series (can be increased though). 

Indeed for historical data, batch methods would be suitable to analyze, and then use something like STUMPI to update when new points arrive. However, I am not sure about the difference between STUMPI, STAMPI and STOMPI yet. Regarding the hardware available, it is CPUs. 

STUMPI can be used to update the matrix profile for a "new" **window** that is arriving. However, since windows are overlapping we can assume the matrix profile is updated once every new point arrives? Regarding the choice for 'm' parameter I am not sure yet (I would need to experiment). 

About deciding whether a new point (or window) is an anomaly based on the matrix profile values, what would you suggest? do you think a percentile approach would be good idea?

**Comment by seanlaw** [comment]
> All of the time series are 1-D and sampled every 10 mins on average (1min-60min). A minimum of 2 weeks of history size is available in memory for each time series (can be increased though). Something like a sliding window where the oldest point is forgotten when a new point arrives.

I think what I'm trying to ascertain is the total number of data points that would exist in a given sliding window. If I understand correctly, if a new data point arrives every 10 minutes (on average), then in 2 weeks you'd have `6 times/hour * 24 hours/day * 14 days = 2016 data points for a two week period`. Does this math look correct? 

Or, for one time series, is there a set of new data data points coming in (instead of a single data point)? If it's a set, how many data points in that new set?

> Indeed for historical data, batch methods would be suitable to analyze, and then use something like STUMPI to update when new points arrive. However, I am not sure about the difference between STUMPI, STAMPI and STOMPI yet. Regarding the hardware available, it is CPUs.

Don't worry about STUMPI/STAMPI/STOMPI. Just know that STUMPI will implement what is published in the STOMPI paper

> However, since windows are overlapping we can assume the matrix profile is updated once every new point arrives? 

You'll likely need to figure out which data points are the "new" points and tell `STUMPI` that those are the points that you'll be appending

> About deciding whether a new point (or window) is an anomaly based on the matrix profile values, what would you suggest? do you think a percentile approach would be good idea?

This is the trickier part. Do you think that you'll maintain the full matrix profile for all of time and then incrementally extend it? Or do you think that you'll only be interested in, say, the last month of data and always get rid of the matrix profile prior to that time? 

Rather than a percentile approach, I prefer a stddev approach. So, you compute the matrix profile, find the standard deviation of the matrix profile and then consider points that are 2 stddevs outside of the mean to be potential anomalies. Note that there is some work being done on this in [this issue](https://github.com/TDAmeritrade/stumpy/issues/184) and [this corresponding PR](https://github.com/TDAmeritrade/stumpy/pull/185)

Obviously, we won't perform that truncation on our end but these are things that you'll need to consider. We'll try to make everything as generalized as possible and make as few assumptions as possible while staying faithful to the paper.

**Comment by hmcoservit** [comment]
Yes on average it would be 2000 points for a 10 minute sampled time series for a 2 week history size. It can be up to around 20000 for time series sampled every minute...

Usually when data points come in, they are as single data points (it may be a set if delays occur in ingestion).

Regarding maintaining the matrix profile (or history size), for example the last month of data can be maintained and we can always get rid of the matrix profile prior to that time when new data points arrive.

And finally thank you for the suggestion about the stddev approach.

**Comment by seanlaw** [comment]
@hmcoservit Thank you for the details. The reason why I wanted to know about the data size is because it is estimated that it should take roughly 2-7 seconds on a 2-core machine (faster when there are more cores) to process between 16,384-32,768 data points using straight `stumpy.stump` (according to some of our very rough timing comparisons - [See this table](https://stumpy.readthedocs.io/en/latest/index.html#performance)). Of course, this is computing the full matrix profile every time.

To be clear, this is not me being against adding STUMPI but I just wanted to make sure that your data size is possibly still in the range of being "fast enough" to reanalyze every time. Certainly, this is not my call.

I'm currently taking a look at this. If I posted some code here, would you be up for testing it out on your end? I'm basically implementing Table 5 from [this paper](https://www.cs.ucr.edu/~eamonn/MP_journal.pdf). I'd appreciate any questions, comments, or concerns that you may have.

**Comment by hmcoservit** [comment]
@seanlaw Thank you for your answer, you are right. Using stump right now for my benchmark tests is probably sufficient to evaluate the accuracy. STUMPI would be more advantageous at a later stage in a big data environment (production) where there are data points from thousands of time series coming in as a stream (sampled differently). In that case speed would be critical. 

For my tests, the first step would be regarding testing the accuracy of matrix profile algorithms for anomaly detection and comparing it to the other unsupervised anomaly detection methods (I have labeled data in hand). The next steps would be regarding speed and scalability where STUMPI would certainly be useful. The final output of the tests would help choose an anomaly detection algorithm (balancing speed, accuracy and scalability) to be used in a big data streaming scenario to do automated anomaly detection in real time for thousand of time series. 

No problem regarding testing the code, I would be more than happy to do that.

**Comment by seanlaw** [comment]
@hmcoservit I think I have a reasonable working prototype but I still need to add unit tests and handling of `np.nan`/`np.inf` values in the time series so that it is consistent with the rest of the code base. Would you mind taking a look at the following, test it out, see if it makes sense, and ask any questions that you may have?

I've tried to keep it as close to the original publication as possible but note that the paper uses 1-based indexing instead of zero-based. I've verified the results by comparing it to the output of `stumpy.stump`:

```
import numpy as np
from stumpy import core
import stumpy
import numpy.testing as npt


def stumpi(t, T, m, P, I, QT, M_T, Σ_T):
    n = T.shape[0]
    l = n - m + 1
    T_new = np.append(T, t)
    QT_new = np.empty(QT.shape[0]+1)
    S = T_new[l:]
    t_drop = T_new[l-1]

    for j in range(l, 0, -1):
        QT_new[j] = QT[j-1] - T_new[j-1] * t_drop + T_new[j + m - 1] * t
    QT_new[0] = 0

    for j in range(m):
        QT_new[0] = QT_new[0] + T_new[j] * S[j]
        
    μ_Q = M_T[l-1] + (t - t_drop) / m
    σ_Q = np.sqrt(Σ_T[l-1] * Σ_T[l-1] + M_T[l-1] * M_T[l-1] + (t * t - t_drop * t_drop) / m - μ_Q * μ_Q)
    
    M_T_new = np.append(M_T, μ_Q)
    Σ_T_new = np.append(Σ_T, σ_Q)
    D = core.calculate_distance_profile(m, QT_new, μ_Q, σ_Q, M_T_new, Σ_T_new)
    
    excl_zone = int(np.ceil(m / 4))
    core.apply_exclusion_zone(D, D.shape[0]-1, excl_zone)
    
    for j in range(l):
        if D[j] < P[j]:
            I[j] = l
            P[j] = D[j]
    
    I_last = np.argmin(D)
    I_new = np.append(I, I_last)
    P_new = np.append(P, D[I_last])
    
    return T_new, P_new, I_new, QT_new, M_T_new, Σ_T_new


if __name__ == "__main__":
    # Create a new time series with 20 elements and set the window size to 4
    T = np.random.rand(20)
    m = 4

    # Obtain the first matrix profile before incrementally streaming a new data point
    T, M_T, Σ_T = core.preprocess(T, m)
    mp = stumpy.stump(T, m)
    P = mp[:, 0].copy()
    I = mp[:, 1].copy()
    Q = T[-m:]
    QT = core.sliding_dot_product(Q, T)

    # Now we stream in 1,000 new data points one at a time and update our matrix profile
    for i in range(1000):
        t = np.random.rand()
        T, P, I, QT, M_T, Σ_T = stumpi(t, T, m, P, I, QT, M_T, Σ_T)

    # For comparison, let's compute the matrix profile using a typical batch method
    mp = stumpy.stump(T, m)

    # And check that the results are identical
    npt.assert_almost_equal(mp[:, 0], P)
    npt.assert_almost_equal(mp[:, 1], I)
```

**Comment by seanlaw** [comment]
@mexxexx Are you able to provide any thoughts or suggestions on what needs to be done in the the code from the last comment so that it would also work for `nan`/`inf`?

**Comment by mexxexx** [comment]
I don't know much about STOMPI, but maybe this helps:
* I guess you recalculate the distance profile of the last window, so of that contains a nan/inf value, the distance profile should be set to inf everywhere .
* The means `M_T` should be inf wherever a sequence contains a nan/inf value.

Everything else (i.e. setting the distance profile to inf in the appropriate positions) should be automatically done by `core.mass` if the definition of `M_T` is consistent with the other algorithms.

**Comment by mexxexx** [comment]
> The one thing that bothers me is having to create a new array every time a new data point comes in. 

I guess you are talking about `QT_new`. How about resizing QT by doubling the length if necessary and modifying it in place?

**Comment by seanlaw** [comment]
> I don't know much about STOMPI, but maybe this helps:
> 
> * I guess you recalculate the distance profile of the last window, so of that contains a nan/inf value, the distance profile should be set to inf everywhere .
> 
> * The means `M_T` should be inf wherever a sequence contains a nan/inf value.
> 
> 
> 
> Everything else (i.e. setting the distance profile to inf in the appropriate positions) should be automatically done by `core.mass` if the definition of `M_T` is consistent with the other algorithms.

Unfortunately, if you look at the STUMPI function above it calls `core.calculate_distance_profile` rather than `core.mass`. That’s because it updates `QT` each time in constant time (since this is faster than performing FFT in `core.mass`) and it also derives and updates `M_T` and `Σ_T` as well. 

Correct me if I’m wrong but we should first convert the `inf` to `nan` before computing `M_T` and `Σ_T` with `core.compute_mean_stddev` (having `nan` in the time series is fine). After this, we should set all `nan` in the time series to zero before computing `QT`. Once `QT` is computed then it can be passed into `core.calculate_distance_profile` without doing anything else to `QT`. I am just thinking out loud here.

**Comment by seanlaw** [comment]
> I guess you are talking about `QT_new`. How about resizing QT by doubling the length if necessary and modifying it in place?

Actually all of the values that it returns that end with `_new` are all “new” arrays that were newly allocated within the function. They are just extended by a single element for each iteration.

I like your idea about doubling the array but I would want a way to limit what the user can see (or inspect in the array) even if the array were much larger than it really is. I wonder if there’s a way to accomplish that in NumPy.

**Comment by hmcoservit** [comment]
@seanlaw Great! thank you for the code, I will test it out as soon as possible and let you know the results. And thanks for pointing out the issue with removal of the first data point, I will keep that in mind.

**Comment by seanlaw** [comment]
Okay, I've decided to move `np.nan`/`np.inf` support to a separate issue as it is a bit trickier to handle in both the pre-processing step as well as the in function step. For now, users just need to ensure that there are no `np.nan`/`np.inf` contained within their time series. Otherwise, it will raise a `ValueError` accordingly.

**Comment by seanlaw** [comment]
@hmcoservit I'm going to put together a [short tutorial for our documentation](https://github.com/TDAmeritrade/stumpy/issues/200) but I've re-opened this issue so feel free to comment on what you think of the implementation above.

**Comment by mexxexx** [comment]
> Unfortunately, if you look at the STUMPI function above it calls core.calculate_distance_profile rather than core.mass. That’s because it updates QT each time in constant time (since this is faster than performing FFT in core.mass) and it also derives/reverse engineers and updates M_T and Σ_T as well.

I see what you mean, however, doesn't `core.calculate_distance_profile` also do exactly the thing I described?

So modifiying this
```
D = core.calculate_distance_profile(m, QT_new, μ_Q, σ_Q, M_T_new, Σ_T_new)
```

to 
```
if np.isnan(μ_Q):
    D = np.full(M_T_new.size, np.inf, dtype=float)
else:
    D = core.calculate_distance_profile(m, QT_new, μ_Q, σ_Q, M_T_new, Σ_T_new)
```
could work. If 'μ_Q' is `inf`, then we set the whole distance profile to `inf`. Otherwise, wherever `M_T_new` is `inf`, we would need to set D to `inf`, and that is exactly what ` core.calculate_distance_profile` is doing (compare it to how we do nan/inf handling in the other algorithms).

The only thing to be careful about is how you compute `μ_Q` and `σ_Q` if `M_T[-1]` is `inf`. Probably the best thing would be to set 
```
μ_Q, σ_Q = core.compute_mean_std(T[n-m+1:], m)
```

**Comment by seanlaw** [comment]
@mexxexx Thank you for your precious insights. Indeed, I did go down the path of using `core.compute_mean_std` but I was still getting stuck and then I wasn't sure what to borrow from other functions. I'll take another stab at it with your suggestions!

**Comment by seanlaw** [comment]
This is now completed and available in v1.4.0.

Thank you all for helping out and chiming in!
